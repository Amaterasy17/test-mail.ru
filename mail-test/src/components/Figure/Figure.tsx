import React, { useEffect, useState } from 'react';
import { Loader } from './Loader/Loader';
import {
	Figcaption,
	FigureContainer,
	FigureDiv,
	Image,
	ImageContainer,
} from './styles';
import { Icon } from './Icon/Icon';

export type FigureProps = {
	children?: React.ReactNode,
	src: string,
	arWidth?: number,
	arHeight?: number,
};

type ChildrenProps = {
	children: React.ReactNodeArray,
};

export const Figure: React.FC<FigureProps> = ({
	children,
	src,
	arWidth = 9,
	arHeight = 16,
}: FigureProps) => {
	const [height, setHeight] = useState<number>(0);
	const [loading, setLoading] = React.useState<boolean>(true);
	const [error, setError] = React.useState<boolean>(false);

	const ComponentContainer = React.forwardRef<HTMLDivElement, ChildrenProps>(
		(props, ref) => (
			<FigureDiv ref={ref} height={height}>
				{props.children}
			</FigureDiv>
		)
	);
	const ref = React.createRef<HTMLDivElement>();

	const onLoading = (ev: React.SyntheticEvent<HTMLImageElement>) => {
		setLoading(false);
	};

	const onError = (ev: React.SyntheticEvent<HTMLImageElement>) => {
		setLoading(false);
		setError(true);
	};

	useEffect(() => {
		if (!ref.current) {
			return;
		}

		const width: number = ref.current.clientWidth;
		const height = (arHeight * width) / arWidth;

		if (height < ref.current.clientHeight) {
			setHeight(height);
		}
	}, []);

	return (
		<ComponentContainer ref={ref}>
			<FigureContainer>
				<ImageContainer arHeight={arHeight} arWidth={arWidth}>
					{loading && <Loader />}
					<Image
						isNone={loading || error}
						src={src}
						alt={'Загрузка'}
						onLoad={onLoading}
						onError={onError}
					/>
					{error && <Icon />}
				</ImageContainer>
			</FigureContainer>
			<Figcaption>{children}</Figcaption>
		</ComponentContainer>
	);
};
