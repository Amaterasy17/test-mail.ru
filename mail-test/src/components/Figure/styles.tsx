import styled, { css } from 'styled-components';

type StyledImgProps = {
	isNone: boolean,
};

type StyledImage = {
	arWidth: number,
	arHeight: number,
};

type StyledDivProps = {
	height: number,
};

export const FigureDiv = styled.div`
	width: 100%;
	height: ${(props: StyledDivProps) => {
		if (props.height === 0) {
			return '100%';
		}
		return `${props.height}px`;
	}};
`;

export const FigureContainer = styled.div`
	display: flex;
	flex-direction: column;
	align-items: center;
	justify-content: center;
	width: 100%;
	box-sizing: border-box;
	height: 100%;
	overflow: hidden;
`;

export const ImageContainer = styled.div`
	display: flex;
	flex-direction: column;
	justify-content: center;
	background-color: #1f2128;
	align-items: center;
	position: relative;
	width: 100%;
	height: 100%;
	padding-top: calc(
		${(props: StyledImage) => props.arHeight / props.arWidth} * 100%
	);
`;

export const Image = styled.img`
	width: 100%;
	height: 100%;
	display: block;
	object-fit: cover;
	position: absolute;
	top: 0;
	left: 0;
	right: 0;
	bottom: 0;
	${({ isNone }: StyledImgProps) =>
		isNone &&
		css`
			display: none;
		`}
`;

export const Figcaption = styled.figcaption`
	font-family: -apple-system, 'Roboto', 'Oxygen', sans-serif;
	font-size: 2.5vh;
	margin-top: 3px;
	text-align: center;
`;
