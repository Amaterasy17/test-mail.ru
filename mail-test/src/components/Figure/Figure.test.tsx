import React from 'react';
import { configure, render } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { Figure } from './Figure';
import styled from 'styled-components';

configure({ adapter: new Adapter() });

type testProps = {
	width: number,
	height: number,
};

const TestDiv = styled.div`
	width: ${(props: testProps) => `${props.width}px`};
	height: ${(props: testProps) => `${props.height}px`};
`;

jest.mock('react', () => ({
	...jest.requireActual('react'),
	useLayoutEffect: jest.requireActual('react').useEffect,
}));

const photoSrc =
	'https://images.pexels.com/photos/1647214/pexels-photo-1647214.jpeg?dl&fit=crop&crop=entropy&w=640&h=1006';

describe('Tests of Figure component', () => {
	afterAll(() => {
		jest.resetModules();
	});

	test.each([
		[300, 500, 5, 3, photoSrc],
		[300, 500, 2, 1, photoSrc],
		[300, 500, 5, 1, photoSrc],
		[300, 100, 16, 9, photoSrc],
		[300, 500, 4, 3, photoSrc],
		[300, 500, 15, 5, photoSrc],
	])(
		'tests of resizing pictures',
		(
			width: number,
			height: number,
			arHeight: number,
			arWidth: number,
			src: string
		) => {
			const figure = render(
				<TestDiv width={width} height={height}>
					<Figure src={src} arWidth={arWidth} arHeight={arHeight} />
				</TestDiv>
			);
			expect(figure).toMatchSnapshot();
		}
	);
	it('test with not valid src of photo', () => {
		const figure = render(
			<TestDiv width={300} height={500}>
				<Figure src="fakePicture" arWidth={9} arHeight={16} />
			</TestDiv>
		);
		expect(figure).toMatchSnapshot();
	});

	it('default width and height with figcaption', () => {
		const figure = render(
			<TestDiv width={300} height={300}>
				<Figure src="https://images.pexels.com/photos/1647214/pexels-photo-1647214.jpeg?dl&fit=crop&crop=entropy&w=640&h=1006">
					Автор:{' '}
					<a href="https://www.pexels.com/@simonmigaj">Simon Migaj</a>
				</Figure>
			</TestDiv>
		);
		expect(figure).toMatchSnapshot();
	});
});
