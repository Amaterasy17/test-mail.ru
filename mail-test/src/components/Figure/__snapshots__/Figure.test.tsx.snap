// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests of Figure component default width and height with figcaption 1`] = `
<div
  class="sc-fujyAs cwgfdV"
  height="300"
  width="300"
>
  <div
    class="sc-gtsrHT jOAJgy"
    height="0"
  >
    <div
      class="sc-dlnjwi kbBcQf"
    >
      <div
        class="sc-hKFxyN bQqzNx"
      >
        <div
          class="loaderRing-0-2-1"
        >
          <div
            class="sc-bdnxRM dxUva loaderRingPart-0-2-2 loaderAnimationRing-0-2-3 loaderAnimationRing-d21-0-2-25"
          />
          <div
            class="sc-bdnxRM dxUva loaderRingPart-0-2-2 loaderAnimationRing-0-2-3 loaderAnimationRing-d22-0-2-26"
          />
          <div
            class="sc-bdnxRM dxUva loaderRingPart-0-2-2 loaderAnimationRing-0-2-3 loaderAnimationRing-d23-0-2-27"
          />
          <div
            class="sc-bdnxRM dxUva"
          />
        </div>
        <img
          alt="Загрузка"
          class="sc-eCApnc dtPAjT"
          src="https://images.pexels.com/photos/1647214/pexels-photo-1647214.jpeg?dl&fit=crop&crop=entropy&w=640&h=1006"
        />
      </div>
    </div>
    <figcaption
      class="sc-jSFjdj gJENt"
    >
      Автор: 
      <a
        href="https://www.pexels.com/@simonmigaj"
      >
        Simon Migaj
      </a>
    </figcaption>
  </div>
</div>
`;

exports[`Tests of Figure component test with not valid src of photo 1`] = `
<div
  class="sc-fujyAs cweGFr"
  height="500"
  width="300"
>
  <div
    class="sc-gtsrHT jOAJgy"
    height="0"
  >
    <div
      class="sc-dlnjwi kbBcQf"
    >
      <div
        class="sc-hKFxyN bQqzNx"
      >
        <div
          class="loaderRing-0-2-1"
        >
          <div
            class="sc-bdnxRM dxUva loaderRingPart-0-2-2 loaderAnimationRing-0-2-3 loaderAnimationRing-d18-0-2-22"
          />
          <div
            class="sc-bdnxRM dxUva loaderRingPart-0-2-2 loaderAnimationRing-0-2-3 loaderAnimationRing-d19-0-2-23"
          />
          <div
            class="sc-bdnxRM dxUva loaderRingPart-0-2-2 loaderAnimationRing-0-2-3 loaderAnimationRing-d20-0-2-24"
          />
          <div
            class="sc-bdnxRM dxUva"
          />
        </div>
        <img
          alt="Загрузка"
          class="sc-eCApnc dtPAjT"
          src="fakePicture"
        />
      </div>
    </div>
    <figcaption
      class="sc-jSFjdj gJENt"
    />
  </div>
</div>
`;

exports[`Tests of Figure component tests of resizing pictures 1`] = `
<div
  class="sc-fujyAs cweGFr"
  height="500"
  width="300"
>
  <div
    class="sc-gtsrHT jOAJgy"
    height="0"
  >
    <div
      class="sc-dlnjwi kbBcQf"
    >
      <div
        class="sc-hKFxyN dgWisQ"
      >
        <div
          class="loaderRing-0-2-1"
        >
          <div
            class="sc-bdnxRM dxUva loaderRingPart-0-2-2 loaderAnimationRing-0-2-3 loaderAnimationRing-d0-0-2-4"
          />
          <div
            class="sc-bdnxRM dxUva loaderRingPart-0-2-2 loaderAnimationRing-0-2-3 loaderAnimationRing-d1-0-2-5"
          />
          <div
            class="sc-bdnxRM dxUva loaderRingPart-0-2-2 loaderAnimationRing-0-2-3 loaderAnimationRing-d2-0-2-6"
          />
          <div
            class="sc-bdnxRM dxUva"
          />
        </div>
        <img
          alt="Загрузка"
          class="sc-eCApnc dtPAjT"
          src="https://images.pexels.com/photos/1647214/pexels-photo-1647214.jpeg?dl&fit=crop&crop=entropy&w=640&h=1006"
        />
      </div>
    </div>
    <figcaption
      class="sc-jSFjdj gJENt"
    />
  </div>
</div>
`;

exports[`Tests of Figure component tests of resizing pictures 2`] = `
<div
  class="sc-fujyAs cweGFr"
  height="500"
  width="300"
>
  <div
    class="sc-gtsrHT jOAJgy"
    height="0"
  >
    <div
      class="sc-dlnjwi kbBcQf"
    >
      <div
        class="sc-hKFxyN bMLozk"
      >
        <div
          class="loaderRing-0-2-1"
        >
          <div
            class="sc-bdnxRM dxUva loaderRingPart-0-2-2 loaderAnimationRing-0-2-3 loaderAnimationRing-d3-0-2-7"
          />
          <div
            class="sc-bdnxRM dxUva loaderRingPart-0-2-2 loaderAnimationRing-0-2-3 loaderAnimationRing-d4-0-2-8"
          />
          <div
            class="sc-bdnxRM dxUva loaderRingPart-0-2-2 loaderAnimationRing-0-2-3 loaderAnimationRing-d5-0-2-9"
          />
          <div
            class="sc-bdnxRM dxUva"
          />
        </div>
        <img
          alt="Загрузка"
          class="sc-eCApnc dtPAjT"
          src="https://images.pexels.com/photos/1647214/pexels-photo-1647214.jpeg?dl&fit=crop&crop=entropy&w=640&h=1006"
        />
      </div>
    </div>
    <figcaption
      class="sc-jSFjdj gJENt"
    />
  </div>
</div>
`;

exports[`Tests of Figure component tests of resizing pictures 3`] = `
<div
  class="sc-fujyAs cweGFr"
  height="500"
  width="300"
>
  <div
    class="sc-gtsrHT jOAJgy"
    height="0"
  >
    <div
      class="sc-dlnjwi kbBcQf"
    >
      <div
        class="sc-hKFxyN btDOcX"
      >
        <div
          class="loaderRing-0-2-1"
        >
          <div
            class="sc-bdnxRM dxUva loaderRingPart-0-2-2 loaderAnimationRing-0-2-3 loaderAnimationRing-d6-0-2-10"
          />
          <div
            class="sc-bdnxRM dxUva loaderRingPart-0-2-2 loaderAnimationRing-0-2-3 loaderAnimationRing-d7-0-2-11"
          />
          <div
            class="sc-bdnxRM dxUva loaderRingPart-0-2-2 loaderAnimationRing-0-2-3 loaderAnimationRing-d8-0-2-12"
          />
          <div
            class="sc-bdnxRM dxUva"
          />
        </div>
        <img
          alt="Загрузка"
          class="sc-eCApnc dtPAjT"
          src="https://images.pexels.com/photos/1647214/pexels-photo-1647214.jpeg?dl&fit=crop&crop=entropy&w=640&h=1006"
        />
      </div>
    </div>
    <figcaption
      class="sc-jSFjdj gJENt"
    />
  </div>
</div>
`;

exports[`Tests of Figure component tests of resizing pictures 4`] = `
<div
  class="sc-fujyAs cwfECj"
  height="100"
  width="300"
>
  <div
    class="sc-gtsrHT jOAJgy"
    height="0"
  >
    <div
      class="sc-dlnjwi kbBcQf"
    >
      <div
        class="sc-hKFxyN bQqzNx"
      >
        <div
          class="loaderRing-0-2-1"
        >
          <div
            class="sc-bdnxRM dxUva loaderRingPart-0-2-2 loaderAnimationRing-0-2-3 loaderAnimationRing-d9-0-2-13"
          />
          <div
            class="sc-bdnxRM dxUva loaderRingPart-0-2-2 loaderAnimationRing-0-2-3 loaderAnimationRing-d10-0-2-14"
          />
          <div
            class="sc-bdnxRM dxUva loaderRingPart-0-2-2 loaderAnimationRing-0-2-3 loaderAnimationRing-d11-0-2-15"
          />
          <div
            class="sc-bdnxRM dxUva"
          />
        </div>
        <img
          alt="Загрузка"
          class="sc-eCApnc dtPAjT"
          src="https://images.pexels.com/photos/1647214/pexels-photo-1647214.jpeg?dl&fit=crop&crop=entropy&w=640&h=1006"
        />
      </div>
    </div>
    <figcaption
      class="sc-jSFjdj gJENt"
    />
  </div>
</div>
`;

exports[`Tests of Figure component tests of resizing pictures 5`] = `
<div
  class="sc-fujyAs cweGFr"
  height="500"
  width="300"
>
  <div
    class="sc-gtsrHT jOAJgy"
    height="0"
  >
    <div
      class="sc-dlnjwi kbBcQf"
    >
      <div
        class="sc-hKFxyN frDKwB"
      >
        <div
          class="loaderRing-0-2-1"
        >
          <div
            class="sc-bdnxRM dxUva loaderRingPart-0-2-2 loaderAnimationRing-0-2-3 loaderAnimationRing-d12-0-2-16"
          />
          <div
            class="sc-bdnxRM dxUva loaderRingPart-0-2-2 loaderAnimationRing-0-2-3 loaderAnimationRing-d13-0-2-17"
          />
          <div
            class="sc-bdnxRM dxUva loaderRingPart-0-2-2 loaderAnimationRing-0-2-3 loaderAnimationRing-d14-0-2-18"
          />
          <div
            class="sc-bdnxRM dxUva"
          />
        </div>
        <img
          alt="Загрузка"
          class="sc-eCApnc dtPAjT"
          src="https://images.pexels.com/photos/1647214/pexels-photo-1647214.jpeg?dl&fit=crop&crop=entropy&w=640&h=1006"
        />
      </div>
    </div>
    <figcaption
      class="sc-jSFjdj gJENt"
    />
  </div>
</div>
`;

exports[`Tests of Figure component tests of resizing pictures 6`] = `
<div
  class="sc-fujyAs cweGFr"
  height="500"
  width="300"
>
  <div
    class="sc-gtsrHT jOAJgy"
    height="0"
  >
    <div
      class="sc-dlnjwi kbBcQf"
    >
      <div
        class="sc-hKFxyN dJXxJ"
      >
        <div
          class="loaderRing-0-2-1"
        >
          <div
            class="sc-bdnxRM dxUva loaderRingPart-0-2-2 loaderAnimationRing-0-2-3 loaderAnimationRing-d15-0-2-19"
          />
          <div
            class="sc-bdnxRM dxUva loaderRingPart-0-2-2 loaderAnimationRing-0-2-3 loaderAnimationRing-d16-0-2-20"
          />
          <div
            class="sc-bdnxRM dxUva loaderRingPart-0-2-2 loaderAnimationRing-0-2-3 loaderAnimationRing-d17-0-2-21"
          />
          <div
            class="sc-bdnxRM dxUva"
          />
        </div>
        <img
          alt="Загрузка"
          class="sc-eCApnc dtPAjT"
          src="https://images.pexels.com/photos/1647214/pexels-photo-1647214.jpeg?dl&fit=crop&crop=entropy&w=640&h=1006"
        />
      </div>
    </div>
    <figcaption
      class="sc-jSFjdj gJENt"
    />
  </div>
</div>
`;
